lib/test.js

```diff
@@ -135,27 +135,38 @@ Test.prototype.run = function () {
 	ret = observableToPromise(ret);
 
 	if (isPromise(ret)) {
-		ret
-			.then(function () {
-				if (!self.planCount || self.planCount === self.assertCount) {
-					self.exit();
-				}
-			})
-			.catch(function (err) {
-				self._setAssertError(new assert.AssertionError({
-					actual: err,
-					message: 'Promise rejected → ' + err,
-					operator: 'promise'
-				}));
-
+		if (!this.metadata.async) {
+			ret = ret.then(function () {
 				self.exit();
 			});
+		}
+
+		ret.catch(function (err) {
+			self._setAssertError(new assert.AssertionError({
+				actual: err,
+				message: 'Promise rejected → ' + err,
+				operator: 'promise'
+			}));
+
+			self.exit();
+		});
+	} else if (!this.metadata.async) {
+		this.exit();
 	}
 
 	return this.promise.promise;
 };
 
-Test.prototype.end = function (err) {
+Object.defineProperty(Test.prototype, 'end', {
+	get: function () {
+		if (this.metadata.async) {
+			return this._end;
+		}
+		throw new Error('t.end is not supported in this context. To use t.end as a callback, you must explicitly declare the test asynchronous via `test.async(testName, fn)` ');
+	}
+});
+
+Test.prototype._end = function (err) {
```

Comment:
You mean just have `t.end` throw when executed?

Because we want to fail fast, and if passed as a callback:

```js
   fs.readFile(path, t.end);
```

We are waiting for the function to actually be called instead of throwing ASAP.